power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp

```diff
@@ -598,17 +598,130 @@ class TapPositionOptimizerImpl<std::tuple<TransformerTypes...>, StateCalculator,
+            if (tap_diff == 0) {
+                tap_changed = false;
+                return tap_pos;
+            }
+            if ((tap_diff == 1 && previous_down) || (tap_diff == -1 && !previous_down)) {
+                set_bs_last_check(true);
+            }
+            tap_changed = true;
+            set_bs_current_tap(tap_pos);
+            return tap_pos;
+        }
+
+      private:
+        IntS lower_bound{-12};      // tap position lower bound
+        IntS upper_bound{12};       // tap position upper bound
+        IntS current{0};            // current tap position
+        bool last_down{false};      // last direction
+        bool last_check{false};     // last run checked
+        bool tap_reverse{false};    // tap range normal or reversed
+        bool inevitable_run{false}; // inevitable run
+
+        inline IntS get_bs_tap_lower_bound() const { return lower_bound; }
+        inline IntS get_bs_tap_upper_bound() const { return upper_bound; }
+        inline bool get_bs_last_check() const { return last_check; }
+        inline bool get_bs_tap_reverse() const { return tap_reverse; }
+
+        inline void set_bs_tap_lower_bound(IntS _lower_bound) { this->lower_bound = _lower_bound; }
+        inline void set_bs_tap_upper_bound(IntS _upper_bound) { this->upper_bound = _upper_bound; }
+        inline void set_bs_last_down(bool _last_down) { this->last_down = _last_down; }
+        inline void set_bs_tap_reverse(bool _tap_reverse) { this->tap_reverse = _tap_reverse; }
+
+        void reset_bs(IntS tap_pos, IntS tap_min, IntS tap_max) {
+            last_down = false;
+            last_check = false;
+            current = tap_pos;
+            inevitable_run = false;
+            lower_bound = std::min(tap_min, tap_max);
+            upper_bound = std::max(tap_min, tap_max);
+            tap_reverse = tap_max < tap_min;
+        }
+
+        void adjust_bs(bool strategy_max = true) {
+            if (get_bs_last_down()) {
+                set_bs_tap_upper_bound(get_bs_current_tap());
+            } else {
+                set_bs_tap_lower_bound(get_bs_current_tap());
+            }
+            if (get_bs_tap_lower_bound() < get_bs_tap_upper_bound()) {
+                const bool prefer_higher = strategy_max ? !get_bs_tap_reverse() : get_bs_tap_reverse();
+                const IntS tap_pos = search_bs(prefer_higher);
+                if (get_bs_current_tap() == tap_pos) {
+                    return;
+                }
+                set_bs_current_tap(tap_pos);
+            }
+        }
+
+        IntS search_bs(bool prefer_higher = true) const {
+            const auto mid_point = get_bs_tap_lower_bound() + (get_bs_tap_upper_bound() - get_bs_tap_lower_bound()) / 2;
+            if ((get_bs_tap_upper_bound() - get_bs_tap_lower_bound()) % 2 != 0 && prefer_higher) {
+                return static_cast<IntS>(mid_point + 1);
+            }
+            return static_cast<IntS>(mid_point);
```

Comment:
the even vs odd part is a little bit confusing. would it make more sense to do it like this?

```suggestion
            auto const primary_bound = prefer_higher ? get_bs_tap_upper_bound() : get_bs_tap_lower_bound();
            auto const secondary_bound = prefer_higher ? get_bs_tap_lower_bound() : get_bs_tap_upper_bound();
            return std::midpoint(primary_bound, secondary_bound);
```

Reply:
But it's essentially the same, yes.

Reply:
comment addressed in the latest commit.

Reply:
On a second thought, I think a comment here (and above) should be added as the original code section carries more about the logic while the newer is a simplified equivalent, but obscured the logic.